### PR TITLE
feat(msteams): message on change

### PR DESCRIPTION
**Pull request by scefali**
This PR adds the capability to update a Teams message when the user makes an action. Additionally, if an alert triggers for an issue that has an assignee, it will show that assignee in the message. The user can unassign an already assigned issue by clicking "Unassign".

We update the message with a `PUT` request sending the full message which means we need to have the event and rule(s) that triggered the alert so we can keep the message consistent. Note that if there have been other state changes that happened between the notification being sent out and the user performing a state change action, we will show those changes as well. It might be confusing to the user but it is going to show up to date information so I think it will be OK.

I decided to put the assignee information above the buttons. If people have other ideas of where or how to show this, I'd be interested. Image:

Screen Shot 2020-08-13 at 3 35 37 PM

Kind of shitty gif:
MsTeams

### CODE REVIEW

File: src/sentry/integrations/msteams/webhook.py [hunk 1]
```diff
@@ -138,8 +145,8 @@ def post(self, request):
         if data["type"] == "message":
             # the only message events we care about are those which
             # are from a user submitting an option on a card, which
-            # will always contain an "actionType" in the data.
-            if not data["value"] or "actionType" not in data["value"]:
+            # will always contain an "payload.actionType" in the data.
+            if not data["value"].get("payload", {}).get("actionType"):
```
Comment (KKaempen) [inline src/sentry/integrations/msteams/webhook.py:149]: If you're changing the latter two to `.get(...)`, why not the first one as well?

**Comment by KKaempen** [comment]
LGTM

File: src/sentry/integrations/msteams/webhook.py [hunk 2]
```diff
@@ -331,11 +346,42 @@ def handle_action_submitted(self, request):
...
             client.send_card(user_conversation_id, card)
             return self.respond(status=201)
 
-        return self.issue_state_change(group, identity, data["value"])
+        # update the state of the issue
+        issue_change_response = self.issue_state_change(group, identity, data["value"])
+
+        # get the rules from the payload
+        rules = []
+        for rule_id in payload["rules"]:
```
Comment (MeredithAnya) [inline src/sentry/integrations/msteams/webhook.py:359]: could we be doing a filter here instead? 

```python
Rule.objects.filter(id__in=payload["rules"])
```

File: src/sentry/integrations/msteams/webhook.py [hunk 3]
```diff
@@ -331,11 +346,42 @@ def handle_action_submitted(self, request):
...
+            try:
+                rule = Rule.objects.get(id=rule_id)
+                rules.append(rule)
+            except Rule.DoesNotExist:
+                # if the rule doesn't exist anymore, we can still update the issue
+                pass
+
+        # pull the event based off our paload
```
Comment (MeredithAnya) [inline src/sentry/integrations/msteams/webhook.py:367]: > paload

lol

**Comment by MeredithAnya** [comment]
I have a couple general comments

1. I think the added functionality to be able to un-resolve and un-assign is nice, but visually it's really hard to tell that the issue has already be 'triaged' or dealt with. I think that is a benefit of having the card look different, because if you have a stream of messages, they will all look very similar. 

    Aside from changing the UI drastically, given time contraints, I would suggest making the assignee name either a different color, or a different font size? The whole "Assigned to ..." line kind of looks like it's just more informational text before the buttons. Is making the button a different color a possibility? Just trying to think of a way to distinguish better between an issue that has been acted on and one that hasn't.

2. `msteams/utils.py` is a pretty big file and the logic can be a little hard to follow, I saw some comments on certain things that might be able to help but I also think a clean up could be worth doing (soon enough that the code is still fresh in the brain) 

3. There are a lot of functions that, to me, are not obvious that `None` is an expected return value. Many of these help functions start with `build` which implies to me that we are using the arguments given to then format them in a specific way. This seems to be the case normally. However, there are then a handful like

    ```python
        build_group_desc(group)
    ```

   that handle other logic and just return `None` if we don't need that thing built. This one was particularly confusing since I thought `desc` meant descending but maybe it means description? As someone who may (probs likely) have to help maintain this once it's live, I think it's worth pointing out (again maybe this could be done in a clean up, not too far from launch).

**Comment by scefali** [comment]
@MeredithAnya Those are very good suggestions. Let me respond to each of them.

1. I have a meeting with John today to go over some design stuff. I will ask him about your ideas for the cards.
2. I have pulled out the card building code into a new file called `card_builder.py`. It's still 500 lines but at least all of it is related code.
3. I changed the name of that function to `build_group_descr` so hopefully, that's more clear. Not sure what to do about the other function names. Maybe we can chat more later.

**Review by MeredithAnya — APPROVED** [review]
